kong/plugins/grpc-gateway/deco.lua

```diff
@@ -51,166 +65,181 @@ local valid_method = {
-  end, true)
+  end
+  
+  local load_json_names_info = function( file, msg, field )
+    if ( field.json_name ~= nil ) then
+      local full_name = msg.full_name .. "." .. field.name
+  
+      info.json_names[ full_name ] = field.json_name
+    end
+  end
+  
+  local grpc = grpc_tools.new()
+
+  grpc:parse_file( filename, load_bindings_info, load_json_names_info )
+  
+  for _, fn in ipairs( cfg.additional_protos or {} ) do
+    -- We are only interested in types here (necessary for handling `google.protobuf.Any` type)
+    grpc:parse_file( fn, nil, load_json_names_info )
+  end
 
-  _proto_info[fname] = info
+  --[[
+    Sort by regex length to avoid situations where shorter regex matches also longer string
+    e.g. ^/v1/kong/([^/]+)$ matches both "/v1/kong/{uuid}" and "/v1/kong/{uuid}:customMethod" 
+    The former would have uuid = "f05c29ef-b6f7-4b3d-85a0-b84b85794fb1:customMethod"
+  ]]
+  sort_endpoints( info.endpoints )
+
+  cache[ filename ] = info
+  
   return info
 end
 
--- return input and output names of the method specified by the url path
--- TODO: memoize
-local function rpc_transcode(method, path, protofile)
-  if not protofile then
-    return nil
-  end
+--- Return input and output names of the method specified by the url path
+-- @param method
+-- @param path
+-- @param cfg
+-- @return
+-- TODO: memoize - still valid?
+local function rpc_transcode( method, path, cfg )
+  local info = get_call_info( cfg )
 
-  local info = get_proto_info(protofile)
-  info = info[method]
-  if not info then
+  local endpoints = info.endpoints[ method ] or nil
+
+  if not endpoints then
     return nil, ("Unknown method %q"):format(method)
   end
-  for _, endpoint in ipairs(info) do
-    local m, err = re_match(path, endpoint.regex, "jo")
+
+  for _, endpoint in ipairs( endpoints ) do
+    local m, err = re_match( path, endpoint.regex, "jo" )
+
     if err then
       return nil, ("Cannot match path %q"):format(err)
     end
+
     if m then
       local vars = {}
-      for i, name in ipairs(endpoint.varnames) do
-        vars[name] = m[i]
+      for i, name in ipairs( endpoint.varnames ) do
+        -- check handling of dotted variables (e.g. entity.uuid)
+        vars[ name ] = m[ i ]
       end
       return endpoint, vars
     end
   end
+
   return nil, ("Unknown path %q"):format(path)
 end
 
-
-function deco.new(method, path, protofile)
-  if not protofile then
-    return nil, "transcoding requests require a .proto file defining the service"
+--- Constructor
+-- @param method
+-- @param path
+-- @param cfg
+function deco.new( method, path, cfg )
+  if not cfg.proto then
+    return nil, "Transcoding requests require a .proto file defining the service"
   end
 
-  local endpoint, vars = rpc_transcode(method, path, protofile)
+  local endpoint, vars = rpc_transcode( method, path, cfg )
 
   if not endpoint then
-    return nil, "failed to transcode .proto file " .. vars
+    return nil, "Failed to transcode .proto file " .. vars
```

Comment:
Usually, the first character in error messages should not be capitalized. Ref: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Reply:
https://github.com/Kong/kong/pull/8297/commits/b473efb330ab6e30083923d4bc3b804773320e06